### ISSUE TITLE
Optimistic Merging Policy

### ISSUE THREAD

**Issue by laughinghan** [mathquill/mathquill#657]
@stufflebear has been pushing for merging PRs more aggressively, merging whenever it's an overall step forwards even if the PR could be better, just open another PR to further improve it if necessary. I think she's right, so I've been looking for how other, more successful open source projects make this work, and found that ZeroMQ actually does this by [formal policy](http://rfc.zeromq.org/spec:42/C4/). Co-creator Pieter Hintjens [believes you should](http://hintjens.com/blog:95):

> Accept patches like Wikipedia accepts edits. Merge first, fix later, and discuss out of band. Do all work on master. Don't make people wait. You'll get consensus, after the fact.

(He's even written [a whole book](https://hintjens.gitbooks.io/social-architecture/content/) philosophizing about this.)

Basically, in their process:
- contributors (anyone) should discuss problems in the issue tracker before submitting patches
- maintainers should merge every "correct patch" that's submitted, not make "value judgements", but on reviewing they may submit follow-up/counter-patches
- no merging your own patches or committing directly, always someone else reviews and merges
- no topic branches, everyone submits PRs from personal forks

Of course, this hinges on what's a "correct patch", rather than a "value judgement":
- "a minimal and accurate answer to exactly one identified and agreed problem"
- APIs must never break and should be extensible and documented
- match style, pass self-test, be original work, commit message etc

Which really isn't very specific. I chatted with a maintainer in IRC, who explained that they merge "almost anything that is presented in good faith". They've even "had debates about whether a patch should have to pass CI tests as a prerequisite before merging"; "often patches are merged by one contributor, then reverted by another contributor who can cite the problem the patch caused and call the revert a solution to that problem".

That's a bit extreme for me, so I've been thinking of specific guidelines for what should and shouldn't be blockers for merging.

Reasons a PR needs to be fixed before merging (or rejected):
- breaks tests, lint, etc
- breaks API, or inextensible API, or will be burdensome to support long-term (e.g. exposes too much internals)
- problem is not demonstrable or testable, or at least discussed and agreed upon
- new feature is too much code weight
    + this in particular matters more to us than ZeroMQ because clientside JS library weight affects pageload time, whereas a serverside C library weight barely affects startup time
- huge diff with little description, and little discussion beforehand in an issue
- big unnecessary files (repo weight is irreversible)
- someotherhow a step backwards that's difficult to reverse

Things that should be counter-patches, and shouldn't block a PR from being merged:
- non-lint style judgements and nits&mdash;merge and submit a counter-patch!
- solutions that are incomplete but don't break anything
- solutions missing testcases&mdash;but add some `// TODO: needs tests` when merging
- solutions somewhat more complex than necessary, especially if described clearly

The bottom line is a important [attitude change](http://sealedabstract.com/rants/conduct-unbecoming-of-a-hacker/):

> When a patch is proposed ... , there should be a presumption that the patch should be merged.

Thoughts? Particularly @stufflebear @mjdesa @jwmerrill

**Comment by laughinghan** [comment]
(Related reading: a blogpost titled [The Pull Request Hack](
http://felixge.de/2013/03/11/the-pull-request-hack.html) suggests:

> Whenever somebody sends you a pull request, give them commit access to your project.

and provides an interesting supporting anecdote, but didn't inform anything specific about my proposal.)

**Comment by jwmerrill** [comment]
This idea makes me nervous because it's generally so much easier and more rewarding to add code to a project than it is to remove it. That goes double for public APIs, where it's pretty painful to remove anything that has ever shipped.

I get the motivation, though. Right now, it seems like there isn't really enough bandwidth to properly evaluate and provide guidance on all the open PRs. So we probably need to get more people to a point where they're qualified to do that work... somehow.

Maybe we should do a dry run, and try to classify the existing PRs into "would merge" and "wouldn't merge" under this policy. It's usually more productive to have conversations about concrete examples.

**Comment by laughinghan** [comment]
Good points. Dry run is a fantastic idea, will do.

Some preliminary thoughts:

- From chat, I think your main concern is that we might "end up swimming in a sea of incoherent options that helped 1 person out in a specific circumstance". Totally reasonable concern, but maybe it's like what they say in lecture halls about how you should totally raise your hand because "if you have a question, five others are probably wondering the same thing"; maybe it's the case that "if you need this functionality so bad you've gone through the effort of submitting a patch, five others probably need it, too" (as argued by [this lengthy blogpost](http://sealedabstract.com/rants/conduct-unbecoming-of-a-hacker/))
- To prevent the-first-API-anyone-thinks-of from becoming set in stone, ZeroMQ actually suggests that new APIs be marked "draft", and transition to "stable" only after seeing real use
     + should we think about how to prevent accidental use of experimental APIs, like by hiding them behind compile-time flags or omitting them from release builds or maybe simply requiring they be prefixed (`.unstableDropEmbedded()`) or something, or is just marking them "experimental" in docs is good enough?
- Removing code can also be difficult in the sense that it may break functionality in unintended ways that aren't immediately noticeable, but this is essentially a quality problem. Good test coverage and code organization can ensure this is no riskier than any other code change. Wikis are actually known for having surprisingly higher quality than projects that are "picky" about contribution, under the right conditions. But maybe allowing untested code would be bad conditions, I'm definitely worried about proposing that.
- While I see what you're saying about adding code/functionality/APIs being more rewarding than removing some, I think that mainly applies to new contributors, I think people who feel ownership over a project totally find reducing code to be rewarding, like by internal refactoring, or unifying fragmented features/APIs into a more general feature/API, or factoring functionality out into separate plugin thingies.

To foster ownership, the ZeroMQ guy promotes contributors to maintainers "rapidly" (when they "get it"). [The Pull Request Hack](http://felixge.de/2013/03/11/the-pull-request-hack.html) takes this to the extreme, it's about how granting commit access to people who submit PRs _without even reviewing the PR_ lead to contributors taking more responsibility for their patch and the project:

> The result was pure magic. Within a few hours the same person who had just submitted a rather mediocre patch, had now fixed things up and committed them.

I'm basically hoping that if we can hammer out specific yet fairly liberal/low-hassle guidelines for any maintainer to merge patches, then not only would you and Michael feel confident merging PRs without my input being a blocker, but we could invite our recent active contributors like Learnosity to do so, too, which would at least be a step towards our contributors feeling more ownership.

**Comment by mjdesa** [comment]
I like the idea. There seems to be a lot of ideas going on though. Whats the first step in getting some of these ideas implemented?

**Comment by laughinghan** [comment]
Good question. I guess zeroth step is to gather some rules with discriminatory power, then first step is to merge and close a bunch of PRs, then second step is to get rid of all the repo branches and start working off of forks, third is to promote some more maintainers?

**Comment by stufflebear** [comment]
I like the overall idea too. The CI stuff @desa has been working on will be helpful towards something like this. We have two automated checks that run on all open PRs now, and those passing should both be prereqs to merging.

The only guideline I think we should be more stringent on is tests. If a feature that should be tested isn't, we should ask the contributor to add a  couple tests. I think it's very rare that tests are added later, unless it's because the lack of test was already causing a problem.

Labelling new features as experimental or unstable is a good way to test new features. I agree that a lot of the things people go to the effort of implementing are likely features that five other people wanted.

I propose we go through and try this out with the currently open PRs. I think we'll rapidly get a good idea of what works and doesn't work. Then after a couple weeks of us doing it, we could make the guidelines into formal rules and promote more maintainers.
